dwarf-rust-experimental/rust/src/main.rs

```diff
@@ -0,0 +1,502 @@
+        }
+      } else {
+        best_scores.insert(line.address, line);
+      }
+    }
+
+    let source_unit = ScoredSourceUnit {
+      name: unit
+        .name
+        .as_ref()
+        .map(|x| x.to_string())
+        .transpose()?
+        .unwrap_or_default()
+        .to_string(),
+      directory: unit
+        .comp_dir
+        .as_ref()
+        .map(|x| x.to_string())
+        .transpose()?
+        .unwrap_or_default()
+        .to_string(),
+      files: scored_source_files.values().cloned().collect(),
+    };
+    if !source_unit.files.is_empty() {
+      scored_source_units.push(source_unit);
+    }
+  }
+
+  // Iterate through the source units and their files to and remove any
+  // files that don't have any lines in the best_scores map.
+  for unit in &mut scored_source_units {
+    for file in &mut unit.files {
+      file.lines.retain(|line| {
+        if let Some(best_line) = best_scores.get(&line.address) {
+          line.id == best_line.id
+        } else {
+          false
+        }
+      });
+    }
+    unit.files.retain(|file| !file.lines.is_empty());
+  }
+  scored_source_units.retain(|unit| !unit.files.is_empty());
+
+  // convert the scored source units to the final format
+  let res: Vec<SourceUnit> = scored_source_units
+    .into_iter()
+    .map(|unit| SourceUnit {
+      name: unit.name,
+      directory: unit.directory,
+      files: unit
+        .files
+        .into_iter()
+        .map(|file| SourceFile {
+          id: file.id,
+          directory: file.directory,
+          file: file.file,
+          language: file.language,
+        })
+        .collect(),
+    })
+    .collect();
+
+  let mut lines: Vec<Vec<u64>> = best_scores
+    .values()
+    .map(|line| vec![line.address, line.file_id as u64, line.line as u64])
+    .collect();
+  lines.sort_by_key(|line| line[0]);
+
+  Ok(SourceResult {
+    units: Some(res),
+    lines: Some(lines),
+    functions: if functions.is_empty() { None } else { Some(functions) },
+    error: None,
+  })
+}
+
+/// Calculate a score for how "good" a source mapping is. Higher scores are better.
+/// This helps determine which mapping to keep when multiple source locations map to the same address.
+fn calculate_mapping_score(
+  directory: &Option<String>,
+  filename: &str,
+  line_number: u64,
+  row: &LineRow,
+) -> u32 {
+  let mut score = 1000u32; // Base score
+
+  if !row.is_stmt() {
+    score = score.saturating_sub(400); // Heavy penalty for non-statement locations
+  }
+
+  if row.prologue_end() || row.epilogue_begin() {
+    score = score.saturating_sub(300); // Heavy penalty for prologue/epilogue code
+  }
+
+  if row.basic_block() {
+    score = score.saturating_sub(200); // Penalty for basic block boundaries
+  }
+
+  // Strategy 1: Prefer user code over library/dependency code
+  if let Some(dir) = directory {
+    if dir.contains("/rustc/") || dir.contains("/rust/deps/") {
+      score -= 300; // Heavily penalize compiler/dependency paths
```

Comment:
Where those numbers are coming from?
How/when they should be changed?

Reply:
Thanks for explaining!

> I think we will need more examples and user feedback to get this exactly right.

How do we get the user feedback?

Reply:
Bug reports? Does not seem great.  The best strategy is likely to add more examples and examine each line mapping to see if it is actually the best.  But this seems very labor intensive.